src/aspire/source/micrograph.py

```diff
@@ -1,16 +1,225 @@
 import logging
+import os
+from glob import glob
+from pathlib import Path
 
 import numpy as np
 
 from aspire.image import Image
-from aspire.source import Simulation
+from aspire.source import ArrayImageSource, Simulation
 from aspire.source.image import _ImageAccessor
 from aspire.utils import Random, grid_2d
 
 logger = logging.getLogger(__name__)
 
 
-class MicrographSimulation:
+class MicrographSource:
+    def __init__(self, micrographs, dtype=None):
+        """
+        Instantiate a `MicrographSource` with `micrographs`.
+
+        :param micrographs: Numpy array (count, L, L)
+            where  `L` represents the size in pixels.
+            Alternatively, `micrographs` may be a path string,
+            or list of path strings.
+        :param dtype: Data type of returned `MicrographSource`.
+            Default of `None` will attempt to infer dtype from the first micrograph.
+            Explicitly setting dtype will attempt a cast when returning micrographs.
+            Currently only `float32` and `float64` are supported.
+            Note, due to limitations of common MRC implementations,
+            saving is limited to single precision.
+        :return: `MicrographSource`
+        """
+
+        if (micrographs is None) or (
+            hasattr(micrographs, "__len__") and len(micrographs) == 0
+        ):
+            raise RuntimeError(
+                f"Must supply non-empty `micrographs` argument, received {micrographs}."
+            )
+
+        # Passed Array
+        if isinstance(micrographs, np.ndarray):
+            # Ensure dimensions
+            if micrographs.ndim == 2:
+                micrographs = micrographs[None, :, :]
```

Comment:
Maybe this is a contrived case, but what happens if we have a rectangular singleton? Should we lift this into it's own check, so the remaining shape checks still happen?

Reply:
yeah, that's fair, I'll just rearrange the branch. np.